test/suites/clustering.sh

```diff
@@ -780,14 +780,44 @@ test_clustering_storage() {
     LXD_DIR="${LXD_TWO_DIR}" lxc start egg
     LXD_DIR="${LXD_ONE_DIR}" lxc stop egg --force
     LXD_DIR="${LXD_ONE_DIR}" lxc delete egg
+  fi
 
+  # If the driver has the same per-node storage pool config (e.g. size), make sure it's included in the
+  # member_config, and actually added to a joining node so we can validate it.
+  if [ "${poolDriver}" = "zfs" ] || [ "${poolDriver}" = "btrfs" ] || [ "${poolDriver}" = "ceph" ]; then
     # Spawn a third node
     setup_clustering_netns 3
     LXD_THREE_DIR=$(mktemp -d -p "${TEST_DIR}" XXX)
     chmod +x "${LXD_THREE_DIR}"
     ns3="${prefix}3"
-    spawn_lxd_and_join_cluster "${ns3}" "${bridge}" "${cert}" 3 1 "${LXD_THREE_DIR}" "${poolDriver}"
+    LXD_NETNS="${ns3}" spawn_lxd "${LXD_THREE_DIR}" false
+
+    key=$(echo "${driver_config}" | cut -d'=' -f1)
+    value=$(echo "${driver_config}" | cut -d'=' -f2-)
+
+    # Set member_config to match `spawn_lxd_and_join_cluster` for 'data' and `driver_config` for 'pool1'.
+    member_config="{\"entity\": \"storage-pool\",\"name\":\"pool1\",\"key\":\"${key}\",\"value\":\"${value}\"}"
+    if [ "${poolDriver}" = "zfs" ] || [ "${poolDriver}" = "btrfs" ]; then
+      member_config="{\"entity\": \"storage-pool\",\"name\":\"data\",\"key\":\"size\",\"value\":\"1GiB\"},${member_config}"
+    fi
+
+    # Manually send the join request.
+    cert=$(sed ':a;N;$!ba;s/\n/\\n/g' "${LXD_ONE_DIR}/cluster.crt")
+    op=$(curl --unix-socket "${LXD_THREE_DIR}/unix.socket" -X PUT "lxd/1.0/cluster" -d "{\"server_name\":\"node3\",\"enabled\":true,\"member_config\":[${member_config}],\"server_address\":\"10.1.1.103:8443\",\"cluster_address\":\"10.1.1.101:8443\",\"cluster_certificate\":\"${cert}\",\"cluster_password\":\"sekret\"}" | jq -r .operation)
+    curl --unix-socket "${LXD_THREE_DIR}/unix.socket" "lxd${op}/wait"
 
+    # Ensure that node-specific config appears on all nodes,
+    # regardless of the pool being created before or after the node joined.
+    for n in node1 node2 node3 ; do
+      LXD_DIR="${LXD_ONE_DIR}" lxc storage get pool1 "${key}" --target "${n}" | grep -q "${value}"
+    done
+  fi
+
+  if [ "${poolDriver}" = "zfs" ] || [ "${poolDriver}" = "btrfs" ]; then
```

Comment:
I see now its because of the conditional at the top, but why is this test only done for zfs, btrfs and ceph?

Reply:
I had mentioned it earlier, it wasn't clear to me which storage backends we actually use for LXD clustering tests, (looking at https://github.com/lxc/lxc-ci) or how they're set up. These three backends seemed straightforward to me so I wrote the test for them.

Reply:
If LVM is straightforward to set up, I could include that too, though I've never actually tried it.

Reply:
Yep, turns out it was rather straightforward (Except for the slight hiccup of `ubuntu:22.04` images not having the right kernel modules for lvm). So now `lvm` is included in the test. 

As for `dir`, I can't seem to find any node-specific configuration that is supposed to be the same across all nodes. Technically `source` could work, but wouldn't be compatible with the test suite because the daemons are run on the same machine, and so the `source` path for each LXD would point to the same thing.

Reply:
This test suite doesn't use lxc-ci (or the new lxd-ci) and runs without snap fwiw.

Reply:
We do have clustering tests for Dir in this suite so presumably it uses different source locations for each member.

Reply:
We use all storage backends for clustering tests .

Reply:
I think for the purposes of what you're testing here what you've added now is great.

Reply:
Ah I see that now. it had slipped my mind that the tests are set up with github workflows now. With this I can get ceph working with the test suite locally too :)